File: problems/992_subarrays_with_k_different_integers.py
class Solution:
    def subarraysWithKDistinct(self, nums: list[int], k: int) -> int:
        total_count = 0
        num_counts: dict[int, int] = {}
        left = 0
        current_count = 0

        for num in nums:
            count = num_counts.get(num, 0)
            if count == 0:
                k -= 1
            num_counts[num] = count + 1

            if k < 0:
                count2 = num_counts[nums[left]]
                if count2 == 1:
                    k += 1
                num_counts[nums[left]] -= 1
                left += 1
                current_count = 0

            if k == 0:
                while num_counts[nums[left]] > 1:
                    num_counts[nums[left]] -= 1
                    current_count += 1
                    left += 1
                total_count += current_count + 1

        return total_count

File: problems/992-subarrays-with-k-different-integers.ts
function subarraysWithKDistinct(nums: number[], k: number): number {
  let totalCount = 0
  let left = 0
  const numCounts = new Map<number, number>()
  let currentCount = 0

  for (const num of nums) {
    const numCount = numCounts.get(num) ?? 0
    if (numCount === 0) {
      k--
    }
    numCounts.set(num, numCount + 1)

    if (k < 0) {
      const num2 = nums[left]
      const num2Count = numCounts.get(num2)!
      if (num2Count === 1) {
        k++
      }
      left++
      numCounts.set(num2, num2Count - 1)
      currentCount = 0
    }

    if (k === 0) {
      while (numCounts.get(nums[left])! > 1) {
        numCounts.set(nums[left], numCounts.get(nums[left])! - 1)
        currentCount++
        left++
      }
      totalCount += currentCount + 1
    }
  }

  return totalCount
}
